source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc

```diff
@@ -179,6 +179,22 @@ Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers
     // We can only insert trailers at the end of data, so keep track of this value
     // until then.
     grpc_status_ = grpcStatusFromHeaders(headers);
+
+    // This is a header-only response, and we should prepend the gRPC frame
+    // header directly.
+    if (end_stream && withhold_grpc_frames_) {
+      Envoy::Buffer::OwnedImpl data;
+      buildGrpcFrameHeader(data, 0);
+      encoder_callbacks_->addEncodedData(data, false);
+
+      // This call exists to ensure that the content-length is set correctly,
+      // regardless of whether we are withholding grpc frames above.
+      headers.setContentLength(Grpc::GRPC_FRAME_HEADER_SIZE);
+
+      // Insert grpc-status trailers to communicate the error code.
+      auto& trailers = encoder_callbacks_->addEncodedTrailers();
```

Comment:
I'm not suggesting you should use that return code, but it does document the fact that you cannot currently add trailers to response with no data.

Reply:
I guess the part of the documentation that was ambiguous to me is whether the documented restriction applies *only* in the case where `ContinueAndDontEndStream` is returned.

What makes me currently believe that it applies only in this case is the use of the words "in this case":
```
  // TODO(yosrym93): Support adding trailers in this case by implementing new functions to inject
  // trailers, similar to the inject data functions.
```

Thus, I have two questions for you:
1. Are you fairly confident that the restriction applies to *all* cases of header-only response, rather than only the case of a header-only response where the data to be added is not known at the time of encodeHeaders?
2. If so, what negative consequences might result from this change?

Reply:
When I run a debug build of Envoy and an upstream responds with an empty GRPC response, the assertion I linked fails. Release builds disable assertions, so it succeeds.

The negative consequences would be related to the case where there are multiple HTTP filters in use or if multiple filters try to inject trailers. Perhaps @yosrym93 can provide more details?

Reply:
@zuercher I also ran a debug build on the most recent iteration of this PR  and experimented using the Ruby repo linked above. The README on that repo has repro instructions.
 * Without the changes in filter_manager, I also get the assertion failure.
 * With the changes in filter_manager, I no longer get the assertion failure.

**If you ran the most recent version, could you please provide repro instructions? Thanks!**

With respect to negative consequences:
 * As far as I can tell, the restriction on multiple failures trying to inject trailers has always existed and is independent of this change.
 * I'm uncertain about the case with multiple HTTP filters in use.

Reply:
I haven't touched this code in a very long time, so my memory is terrible. I **think** the comment about missing support for injecting trailers is particularly about the case when the trailers will be injected asynchronously (i.e. not during headers iteration).

I am not saying that adding trailers is supported during headers iterations, I can't remember tbh, but that comment was in the context of injecting trailers after headers iteration.

Reply:
Thanks. I'm experimenting with a change that would trigger the body and trailer injection using delayed dispatch outside the encodeHeaders call. I think that might work.

Alternatively, maybe @alyssawilk can say whether it's safe to invoke `addEncodedTrailers` from `encodeHeaders` to set a trailer when we know there's no body. In that case, we'd need to think about what to do with [this assertion](https://github.com/envoyproxy/envoy/blob/main/source/common/http/filter_manager.cc#L1323).

Reply:
I think your assertion link might be incorrect, but if you're referring to the assertion inside `addEncodedTrailers`, this PR's changes in `filter_manager.h` already get around that assertion by setting the LastDataFrame bit.

Reply:
Yeah, I just don't know if subverting that assertion makes sense or not.

Reply:
I think another integration test (see test/integration/filter_integration_test.cc) is in order. Looks like you could add a parameter to the existing test-only [AddBodyFilter filter](https://github.com/envoyproxy/envoy/blob/main/test/integration/filters/add_body_filter.cc) that controls whether it adds trailers in the header and then a test case to exercise it.

Reply:
(Parameters are defined in the matching add_body_filter.proto, I meant to add.)

Reply:
Also, we need to implement this for the decode path as well, and will need a test there too.

Reply:
I can look into adding the test.

For the decode path, are you asking for a change in this filter or a change somewhere else?
I'm asking because I'm not sure what the use case is in this filter, given that well-formed gRPC requests should never have an empty body.

Reply:
It's not necessary for the grpc filter, but the decode (downstream request)  and encode (upstream response) should be mostly symmetric. Need to allow empty requests to have body and trailers added on the decode path.

Reply:
Is there something special I need to do to get `bazel` to pick up the updated proto?

I added a field in `test/integration/filters/add_body_filter.proto`

```
bool add_trailers_in_encode_decode_header = 4;
```

But getting compiler errors when trying to build:

```
bazel test --sandbox_debug :filter_integration_test
test/integration/filters/add_body_filter.cc: In member function 'virtual Envoy::Http::FilterHeadersStatus Envoy::AddBodyStreamFilter::encodeHeaders(Envoy::Http::ResponseHeaderMap&, bool)':
test/integration/filters/add_body_filter.cc:109:22: error: 'using element_type = class Envoy::AddBodyFilterConfig' {aka 'class Envoy::AddBodyFilterConfig'} has no member named 'add_trailers_in_encode_decode_header_'
  109 |         if (config_->add_trailers_in_encode_decode_header_) {
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```
I tried both the spelling with the trailing `_` and without.

Reply:
@zuercher I have added the requested tests, as well as the symmetric change on the decoder side.

Reply:
@zuercher Are there additional changes you'd like to see in this PR, or tests that you'd like to have run?

I haven't written an integration test before in Envoy, so if this was not what you had in mind, please do let me know. :)